lib/Validation.groovy

```diff
@@ -0,0 +1,105 @@
+@Grab(group='org.everit.json', module='org.everit.json.schema', version='1.5.1')
+@Grab(group='org.json', module='json', version='20201115')
+import org.everit.json.schema.Schema
+import org.everit.json.schema.loader.SchemaLoader
+import org.everit.json.schema.ValidationException
+import org.json.JSONObject
+import org.json.JSONTokener
+import org.json.JSONArray
+import groovy.json.JsonSlurper
+import groovy.json.JsonBuilder
+
+class Validation {
+
+    /*
+    * Function to loop over all parameters defined in schema and check
+    * whether the given paremeters adhere to the specificiations
+    */
+    /* groovylint-disable-next-line UnusedPrivateMethodParameter */
+    private static void validateParameters(params, jsonSchema, log) {
+        //=====================================================================//
+        // Validate parameters against the schema
+        InputStream inputStream = new File(jsonSchema).newInputStream()
+        JSONObject rawSchema = new JSONObject(new JSONTokener(inputStream))
+        Schema schema = SchemaLoader.load(rawSchema)
+
+        // Clean the parameters
+        def cleanedParams = cleanParameters(params)
+
+        // Convert to JSONObject
+        def jsonParams = new JsonBuilder(cleanedParams)
+        JSONObject paramsJSON = new JSONObject(jsonParams.toString())
+
+        // Validate
+        try {
+            schema.validate(paramsJSON)
+        } catch (ValidationException e) {
+            log.error 'Found parameter violations!'
+            JSONObject exceptionJSON = e.toJSON()
+            printExceptions(exceptionJSON, log)
+        }
+
+        // Check for nextflow core params and unexpected params
+        def json = new File(jsonSchema).text
+        def Map schemaParams = (Map) new JsonSlurper().parseText(json).get('definitions')
+        def specified_param_keys = params.keySet()
+        def nf_params = ['profile', 'config', 'c', 'C', 'syslog', 'd', 'dockerize',
+                        'bg', 'h', 'log', 'quiet', 'q', 'v', 'version']
+
+        // Collect expected parameters from the schema
+        def expectedParams = []
+        for (group in schemaParams) {
+            for (p in group.value['properties']) {
+                expectedParams.push(p.key)
+            }
+        }
+
+        for (specified_param in specified_param_keys) {
+            // nextflow params
+            if (nf_params.contains(specified_param)) {
+                log.error "ERROR: You have overwritten the core Nextflow parameter -${specified_param} with --${specified_param}!"
+                System.exit(0)
+            }
+            // unexpected params
+            if (!expectedParams.contains(specified_param)) {
+                log.warn "Unexpected parameter specified: ${specified_param}"
```

Comment:
Can we collect these in a map which is available to the rest of the workflow? Would be nice to repeat the log if the workflow exits with an error.

Reply:
Good idea, thanks!